File: frontend/src/shared/api/gitleaks.ts
/**
 * Gitleaks API Client
 */

import { apiClient } from "@/shared/api/serverClient";

export interface GitleaksScanTask {
    id: string;
    project_id: string;
    name: string;
    status: string;
    target_path: string;
    no_git: string;
    total_findings: number;
    scan_duration_ms: number;
    files_scanned: number;
    error_message?: string | null;
    created_at: string;
    updated_at?: string | null;
}

export interface GitleaksFinding {
    id: string;
    scan_task_id: string;
    rule_id: string;
    description?: string | null;
    file_path: string;
    start_line?: number | null;
    end_line?: number | null;
    secret?: string | null;
    match?: string | null;
    commit?: string | null;
    author?: string | null;
    email?: string | null;
    date?: string | null;
    fingerprint?: string | null;
    status: string;
}

export interface GitleaksRule {
    id: string;
    name: string;
    description?: string | null;
    rule_id: string;
    secret_group: number;
    regex: string;
    keywords: string[];
    path?: string | null;
    tags: string[];
    entropy?: number | null;
    is_active: boolean;
    source: string;
    created_at: string;
    updated_at?: string | null;
}

export async function createGitleaksScanTask(params: {
    project_id: string;
    name?: string;
    target_path?: string;
    no_git?: boolean;
}): Promise<GitleaksScanTask> {
    const response = await apiClient.post(`/static-tasks/gitleaks/scan`, params);
    return response.data;
}

export async function getGitleaksScanTask(
    taskId: string,
): Promise<GitleaksScanTask> {
    const response = await apiClient.get(`/static-tasks/gitleaks/tasks/${taskId}`);
    return response.data;
}

export async function interruptGitleaksScanTask(
    taskId: string,
): Promise<{ message: string; task_id: string; status: string }> {
    const response = await apiClient.post(
        `/static-tasks/gitleaks/tasks/${taskId}/interrupt`,
    );
    return response.data;
}

export async function getGitleaksScanTasks(params?: {
    projectId?: string;
    skip?: number;
    limit?: number;
}): Promise<GitleaksScanTask[]> {
    const searchParams = new URLSearchParams();
    if (params?.projectId) searchParams.set("project_id", params.projectId);
    if (params?.skip !== undefined)
        searchParams.set("skip", String(params.skip));
    if (params?.limit !== undefined)
        searchParams.set("limit", String(params.limit));
    const query = searchParams.toString();
    const response = await apiClient.get(
        `/static-tasks/gitleaks/tasks${query ? `?${query}` : ""}`,
    );
    return response.data;
}

export async function getGitleaksFindings(params: {
    taskId: string;
    status?: string;
    skip?: number;
    limit?: number;
}): Promise<GitleaksFinding[]> {
    const searchParams = new URLSearchParams();
    if (params.status) searchParams.set("status", params.status);
    if (params.skip !== undefined) searchParams.set("skip", String(params.skip));
    if (params.limit !== undefined) searchParams.set("limit", String(params.limit));
    const query = searchParams.toString();
    const response = await apiClient.get(
        `/static-tasks/gitleaks/tasks/${params.taskId}/findings${query ? `?${query}` : ""}`,
    );
    return response.data;
}

export async function getGitleaksFinding(params: {
    taskId: string;
    findingId: string;
}): Promise<GitleaksFinding> {
    const response = await apiClient.get(
        `/static-tasks/gitleaks/tasks/${params.taskId}/findings/${params.findingId}`,
    );
    return response.data;
}

export async function updateGitleaksFindingStatus(params: {
    findingId: string;
    status: "open" | "verified" | "false_positive";
}): Promise<{ message: string; finding_id: string; status: string }> {
    const response = await apiClient.post(
        `/static-tasks/gitleaks/findings/${params.findingId}/status`,
        undefined,
        { params: { status: params.status } },
    );
    return response.data;
}

export async function getGitleaksRules(params?: {
    is_active?: boolean;
    source?: string;
    keyword?: string;
    tag?: string;
    skip?: number;
    limit?: number;
}): Promise<GitleaksRule[]> {
    const searchParams = new URLSearchParams();
    if (params?.is_active !== undefined) searchParams.set("is_active", String(params.is_active));
    if (params?.source) searchParams.set("source", params.source);
    if (params?.keyword) searchParams.set("keyword", params.keyword);
    if (params?.tag) searchParams.set("tag", params.tag);
    if (params?.skip !== undefined) searchParams.set("skip", String(params.skip));
    searchParams.set("limit", String(params?.limit ?? 1000));
    const query = searchParams.toString();
    const response = await apiClient.get(`/static-tasks/gitleaks/rules${query ? `?${query}` : ""}`);
    return response.data;
}

export async function getGitleaksRule(ruleId: string): Promise<GitleaksRule> {
    const response = await apiClient.get(`/static-tasks/gitleaks/rules/${ruleId}`);
    return response.data;
}

export async function createGitleaksRule(params: {
    name: string;
    description?: string;
    rule_id: string;
    secret_group?: number;
    regex: string;
    keywords?: string[];
    path?: string;
    tags?: string[];
    entropy?: number;
    is_active?: boolean;
    source?: string;
}): Promise<GitleaksRule> {
    const response = await apiClient.post(`/static-tasks/gitleaks/rules`, params);
    return response.data;
}

export async function updateGitleaksRule(
    ruleId: string,
    params: {
        name?: string;
        description?: string;
        rule_id?: string;
        secret_group?: number;
        regex?: string;
        keywords?: string[];
        path?: string;
        tags?: string[];
        entropy?: number;
        is_active?: boolean;
        source?: string;
    },
): Promise<GitleaksRule> {
    const response = await apiClient.patch(`/static-tasks/gitleaks/rules/${ruleId}`, params);
    return response.data;
}

export async function deleteGitleaksRule(
    ruleId: string,
): Promise<{ message: string; rule_id: string }> {
    const response = await apiClient.delete(`/static-tasks/gitleaks/rules/${ruleId}`);
    return response.data;
}

export async function batchUpdateGitleaksRules(params: {
    rule_ids?: string[];
    source?: string;
    keyword?: string;
    current_is_active?: boolean;
    is_active: boolean;
}): Promise<{ message: string; updated_count: number; is_active: boolean }> {
    const response = await apiClient.post(`/static-tasks/gitleaks/rules/select`, params);
    return response.data;
}

export async function toggleGitleaksRule(
    rule: GitleaksRule,
): Promise<GitleaksRule> {
    return updateGitleaksRule(rule.id, { is_active: !rule.is_active });
}

File: backend_old/app/models/gitleaks.py
"""
Gitleaks 密钥泄露检测模型 - 数据库表定义
包括扫描任务和发现的密钥泄露结果
"""

import uuid
from datetime import datetime
from typing import Optional, List
from sqlalchemy import (
    Column,
    String,
    Integer,
    Float,
    Boolean,
    Text,
    DateTime,
    ForeignKey,
    JSON,
    Index,
    UniqueConstraint,
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.models.base import Base


class GitleaksScanTask(Base):
    """Gitleaks 扫描任务"""
    __tablename__ = "gitleaks_scan_tasks"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(String, ForeignKey("projects.id"), nullable=False)
    name = Column(String, nullable=False)
    status = Column(String, default="pending", comment="pending, running, completed, failed")
    target_path = Column(String, nullable=False, comment="扫描的目标路径")
    no_git = Column(String, default="true", comment="是否不使用 git history")
    total_findings = Column(Integer, default=0, comment="发现的密钥泄露总数")
    scan_duration_ms = Column(Integer, default=0, comment="扫描耗时(毫秒)")
    files_scanned = Column(Integer, default=0, comment="已扫描文件数")
    error_message = Column(Text, nullable=True, comment="错误信息")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    __table_args__ = (
        Index("ix_gitleaks_tasks_project_created_at", "project_id", created_at.desc()),
        Index(
            "ix_gitleaks_tasks_project_lower_status_created_at",
            "project_id",
            func.lower(status),
            created_at.desc(),
        ),
    )

    # Relationships
    project = relationship("Project", back_populates="gitleaks_scan_tasks")
    findings = relationship(
        "GitleaksFinding", back_populates="scan_task", cascade="all, delete-orphan"
    )


class GitleaksFinding(Base):
    """Gitleaks 发现的密钥泄露"""
    __tablename__ = "gitleaks_findings"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    scan_task_id = Column(String, ForeignKey("gitleaks_scan_tasks.id"), nullable=False)
    rule_id = Column(String, nullable=False, comment="规则 ID (如 aws-access-token)")
    description = Column(Text, nullable=True, comment="密钥描述")
    file_path = Column(String, nullable=False, comment="文件路径")
    start_line = Column(Integer, nullable=True, comment="起始行号")
    end_line = Column(Integer, nullable=True, comment="结束行号")
    secret = Column(Text, nullable=True, comment="密钥内容(已脱敏)")
    match = Column(Text, nullable=True, comment="匹配的完整文本")
    commit = Column(String, nullable=True, comment="Git commit SHA (如果有)")
    author = Column(String, nullable=True, comment="作者 (如果有)")
    email = Column(String, nullable=True, comment="邮箱 (如果有)")
    date = Column(String, nullable=True, comment="日期 (如果有)")
    fingerprint = Column(String, nullable=True, comment="Gitleaks 指纹")
    status = Column(String, default="open", comment="open, verified, false_positive")
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    __table_args__ = (
        Index(
            "ix_gitleaks_findings_scan_task_status_created",
            "scan_task_id",
            "status",
            created_at.desc(),
        ),
        Index(
            "ix_gitleaks_findings_scan_task_file_line",
            "scan_task_id",
            "file_path",
            "start_line",
        ),
        Index("ix_gitleaks_findings_fingerprint", "fingerprint"),
    )

    # Relationships
    scan_task = relationship("GitleaksScanTask", back_populates="findings")


class GitleaksRule(Base):
    """Gitleaks 规则定义（结构化字段，运行时渲染为 TOML）"""

    __tablename__ = "gitleaks_rules"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    rule_id = Column(String, nullable=False, comment="gitleaks 规则 ID")
    secret_group = Column(Integer, nullable=False, default=0)
    regex = Column(Text, nullable=False)
    keywords = Column(JSON, nullable=False, default=list)
    path = Column(Text, nullable=True)
    tags = Column(JSON, nullable=False, default=list)
    entropy = Column(Float, nullable=True)
    is_active = Column(Boolean, nullable=False, default=True)
    source = Column(String, nullable=False, default="custom")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    __table_args__ = (
        UniqueConstraint("name", name="uq_gitleaks_rules_name"),
        UniqueConstraint("rule_id", name="uq_gitleaks_rules_rule_id"),
        Index("ix_gitleaks_rules_is_active", "is_active"),
        Index("ix_gitleaks_rules_rule_id", "rule_id"),
        Index("ix_gitleaks_rules_source_active", "source", "is_active"),
    )
